File: Problems/0881. (M) Boats to Save People/Python3/solution.py
class Solution:
  def numRescueBoats(self, people: List[int], limit: int) -> int:
    # sort, then two pointers

    people.sort()
    lo, hi = 0, len(people)-1
    ret = 0
    while lo <= hi:
      # if two people cannot fit, always optimal to ship heaviest person
      if people[lo]+people[hi] <= limit:
        lo += 1
      hi -= 1
      ret += 1

    return ret

File: Problems/0881. (M) Boats to Save People/Java/solution.java
class Solution {
  public int numRescueBoats(int[] people, int limit) {
    // sort, than two pointers

    Arrays.sort(people);
    int lo = 0;
    int hi = people.length - 1;
    int ret = 0;
    while (lo <= hi) {
      // if two people cannot fit, always optimal to ship heaviest person
      if (people[lo]+people[hi] <= limit)
        lo++;
      hi--;
      ret++;
    }
    return ret;
  }
}
